src/common/src/hash/consistent_hash/vnode.rs

```diff
@@ -31,44 +34,82 @@ pub struct VirtualNode(VirtualNodeInner);
 
 /// The internal representation of a virtual node id.
 type VirtualNodeInner = u16;
-static_assertions::const_assert!(VirtualNodeInner::BITS >= VirtualNode::BITS as u32);
+// static_assertions::const_assert!(VirtualNodeInner::BITS >= VirtualNode::BITS as u32);
 
 impl From<Crc32HashCode> for VirtualNode {
     fn from(hash_code: Crc32HashCode) -> Self {
         // Take the least significant bits of the hash code.
         // TODO: should we use the most significant bits?
-        let inner = (hash_code.value() % Self::COUNT as u64) as VirtualNodeInner;
+        let inner = (hash_code.value() % Self::count() as u64) as VirtualNodeInner;
         VirtualNode(inner)
     }
 }
 
 impl VirtualNode {
-    /// The number of bits used to represent a virtual node.
-    ///
-    /// Note: Not all bits of the inner representation are used. One should rely on this constant
-    /// to determine the count of virtual nodes.
-    pub const BITS: usize = 8;
-    /// The total count of virtual nodes.
-    pub const COUNT: usize = 1 << Self::BITS;
+    /// We may use `VirtualNode` as a datum in a stream, or store it as a column.
+    /// Hence this reifies it as a RW datatype.
+    pub const RW_TYPE: DataType = DataType::Int16;
     /// The size of a virtual node in bytes, in memory or serialized representation.
     pub const SIZE: usize = std::mem::size_of::<Self>();
+    /// The minimum (zero) value of the virtual node.
+    pub const ZERO: VirtualNode = unsafe { VirtualNode::from_index_unchecked(0) };
+}
+
+impl VirtualNode {
+    /// The default count of virtual nodes.
+    const DEFAULT_COUNT: usize = 1 << 8;
+    /// The maximum count of virtual nodes, limited by the size of the inner type [`VirtualNodeInner`].
+    const MAX_COUNT: usize = 1 << VirtualNodeInner::BITS;
+
+    /// The total count of virtual nodes.
+    ///
+    /// It can be customized by the environment variable `RW_VNODE_COUNT`, or defaults to [`Self::DEFAULT_COUNT`].
+    pub fn count() -> usize {
+        // Cache the value to avoid repeated env lookups and parsing.
+        static COUNT: LazyLock<usize> = LazyLock::new(|| {
+            if let Ok(count) = std::env::var("RW_VNODE_COUNT") {
```

Comment:
What if the `RW_VNODE_COUNT` is different after restart? 

Considering the vnode is persisted in storage, this will certainly cause some problems, I think. If so, I would question whether it's a good idea to pass it via env var, perhaps making it a MV argument would be better? Or at least it should be a system parameter and we need use the first-time initialized value, like `state_store_url`

Reply:
You are right.

I had some discussion about these approaches in the original issue. Basically this PR is a quick hack if we want to unblock some users to utilize more resources for their clusters. However, if we find this niche and not that urgent, I agree that we may not merge this PR but directly implement the polished ideas, like per-job configuration or global immutable system parameter.